### PR TITLE
feat(omnidreams): add layer-wise DiT parameter offload

### PR DESCRIPTION
**Pull request by Shn0930**
## Purpose

  Add an opt-in layer-wise CPU parameter offload path for the OmniDreams DiT to reduce GPU memory usage during inference.

  The regular OmniDreams path keeps all DiT block parameters resident on CUDA. With layer-wise offload enabled, block parameters remain in pinned CPU memory, while only the current and next blocks are materialized through two reusable CUDA
  staging slots.

  Existing configurations remain unchanged because enable_layerwise_offload defaults to false. This is primarily a GPU-memory optimization rather than a throughput optimization.

  ## Scope

  - Add a reusable LayerwiseOffloader under FlashDreams infrastructure:
      - pack layer parameters into aligned, dtype-specific pinned CPU buffers;
      - preserve Parameter identities while inactive parameters use empty placeholders;
      - use two reusable CUDA staging slots, a dedicated copy stream, and CUDA events;
      - prefetch the next layer while the current layer executes;
      - release materialized layers safely after normal execution or exceptions.

  - Add CosmosTransformerConfig.enable_layerwise_offload, defaulting to false.
  - Enable offload after checkpoint loading and load-time weight transformations.
  - Materialize OmniDreams blocks during:
      - KV-cache initialization;
      - text-embedding KV replacement;
      - regular DiT forward execution.

  - Bypass whole-DiT torch.compile and CUDA Graph execution while offload is active.
  - Reject unsupported combinations and states, including:
      - native DiT acceleration;
      - optimized self- or cross-attention backends;
      - runtime text-edit LoRA;
      - training or gradient-enabled execution;
      - unsupported parameter storage layouts;
      - post-construction dtype/device conversion;
      - state-dict operations after offload is enabled.

  - Extend the existing OmniDreams full-pipeline benchmark with matched eager-resident and layer-wise-offload cases.
  - Add core offloader tests and OmniDreams integration/parity tests.
  - Document configuration and limitations in integrations_v2/omnidreams/README.md.

  No named pipeline preset, Interactive Drive option, or application entry point is added. Existing applications therefore continue to use resident parameters by default.

  ## Test Plan

  Run targeted core and OmniDreams tests:

  .venv/bin/pytest -q \
    flashdreams/tests/test_layerwise_offload.py \
    integrations_v2/omnidreams/tests/test_layerwise_offload.py

  Run CPU regression tests:

  .venv/bin/pytest -m ci_cpu -q --ignore=artifacts

  Run repository checks:

  .venv/bin/pre-commit run --files \
    flashdreams/flashdreams/infra/acceleration/layerwise_offload.py \
    flashdreams/tests/test_layerwise_offload.py \
    integrations_v2/omnidreams/impl/transformer/__init__.py \
    integrations_v2/omnidreams/impl/transformer/network.py \
    integrations_v2/omnidreams/tests/test_layerwise_offload.py \
    integrations_v2/omnidreams/benchmarks/test_pipeline.py \
    integrations_v2/omnidreams/README.md

  Run the GPU comparison in separate fresh processes:

  CUDA_VISIBLE_DEVICES=0 .venv/bin/pytest -q \
    'integrations_v2/omnidreams/benchmarks/test_pipeline.py::test_full_pipeline_layerwise_offload_benchmark[eager-resident]' \
    -p no:manual_marker -m manual --benchmark-only \
    --benchmark-json=/tmp/flashdreams-eager-resident.json

  CUDA_VISIBLE_DEVICES=0 .venv/bin/pytest -q \
    'integrations_v2/omnidreams/benchmarks/test_pipeline.py::test_full_pipeline_layerwise_offload_benchmark[layerwise-offload]' \
    -p no:manual_marker -m manual --benchmark-only \
    --benchmark-json=/tmp/flashdreams-layerwise-offload.json

  The benchmark uses five warmup rounds and 50 measured rounds. The timed region covers generate() only; finalize() runs during teardown and is excluded from latency measurements.

  ## Test Result

  ### Functional validation

  - Targeted core and OmniDreams tests: 26 passed in 5.34s.
  - CUDA two-slot prefetch/event stress test: passed for 100 iterations.
  - Resident/offloaded forward parity: passed.
  - Multi-view execution, cache initialization, and text-KV replacement tests: passed.
  - README configuration example: validated without model instantiation.
  - Ruff, formatting, ty, pre-commit, and git diff --check: passed.

  ### CPU regression

  1962 passed, 2 skipped, 1 failed

  The single failure was:

  flashdreams/tests/test_benchmark_harness.py::
  test_pai_bench_profile_normalizes_public_evaluator_output

  It reproduces independently outside the changed code paths. The test-generated /usr/bin/env python3 wrapper resolves to a system Python environment without numpy, causing the PAI-Bench compatibility import to fail.

  ### GPU benchmark

  Environment and case:

  - NVIDIA H20-3e
  - NVIDIA driver 580.95.05
  - PyTorch 2.12.1+cu130
  - CUDA 13.0
  - cuDNN 9.2.0
  - OmniDreams single-view 2B BF16 checkpoint
  - batch size 1, one view
  - 704x1280 output
  - seed 0
  - compile and CUDA Graphs disabled for both modes
  - five warmup rounds and 50 measured rounds
  - separate fresh processes
  - steady-state generate() timing only

   Mode                  Median generate           p90    Steady CUDA allocated    Peak CUDA allocated
  ━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━━━
   Eager resident             1.080633 s    1.081413 s             9.410452 GiB           9.835384 GiB
  ────────────────────  ─────────────────  ────────────  ───────────────────────  ─────────────────────
   Layer-wise offload         1.082293 s    1.082796 s             6.038056 GiB           6.463369 GiB

  On this measured stack, layer-wise offload:

  - reduced peak allocated GPU memory by 3.3720 GiB (34.28%);
  - reduced steady allocated GPU memory by 3.3724 GiB (35.84%);
  - changed median generate() latency by +0.154%;
  - changed p90 latency by +0.128%;
  - used 3.6094 GiB of pinned host memory;

### CODE REVIEW

**Comment by Shn0930** [comment]
### Layer-wise offload design

  Layer-wise offload is an opt-in, construction-time memory optimization for the OmniDreams DiT.

  After checkpoint loading and load-time weight transforms, each DiT block's parameters are packed by dtype into aligned, pinned CPU master buffers. Existing Parameter objects are preserved, but inactive parameters point to empty placeholders.
  Consequently, moving the pipeline to CUDA does not make every DiT block resident.

  On the first CUDA execution, the offloader allocates two persistent device slots, sized for the largest packed block. A dedicated CUDA copy stream transfers block i + 1 while the current stream executes block i. CUDA events synchronize weight
  readiness and safe slot reuse.

  sequenceDiagram
      participant H as Pinned CPU masters
      participant X as CUDA copy stream
      participant A as GPU slot A
      participant B as GPU slot B
      participant C as Compute stream

      Note over H: One authoritative master per block and dtype
      Note over A,B: Two reusable buffers sized for the largest block
      Note over C: Non-block weights and KV caches remain resident

      H->>X: Materialize block i
      X->>A: Async H2D and bind parameter views
      A-->>C: Wait for ready event

      par Execute current block
          C->>C: Run block i from slot A
      and Prefetch successor
          H->>X: Prefetch block i + 1
          X->>B: Async H2D and bind parameter views
      end

      C-->>A: Record reusable event
      C->>A: Release block i to empty placeholders
      Note over A,B: Next iteration swaps the slots

  Only network.blocks parameters are offloaded. Embedding, projection, and final-layer weights, encoders and decoders, buffers, activations, and attention KV caches remain CUDA-resident.

  Releasing a block does not copy data back to the CPU. The pinned CPU master is authoritative and immutable; the GPU parameter views are simply discarded and rebound to empty placeholders. This is why runtime weight mutation, text-edit LoRA,
  drift correction, training, post-construction dtype conversion, and state-dict operations are incompatible.

  Because block parameter storage changes during execution, whole-DiT torch.compile and CUDA Graph execution are bypassed. The final block prefetches block 0 so the next DiT invocation can begin without an avoidable initial transfer bubble.

File: integrations_v2/omnidreams/impl/transformer/__init__.py [hunk 1]
```diff
@@ -336,6 +342,24 @@ def __init__(self, config: CosmosTransformerConfig) -> None:
                 "Cache-relative RoPE is not supported by native DiT acceleration; "
                 "set native_dit_acceleration='disabled'"
             )
+        if config.enable_layerwise_offload:
```
[the remark this answers is not in the document]
Comment (Shn0930) [reply integrations_v2/omnidreams/impl/transformer/__init__.py:345]: Resolved in ef206aed. We chose to reject the incompatible combination rather than mutate the offloader's CPU masters. Crazy Robotaxi now rejects layer-wise offload with either fused or unfused drift correction before constructing the corrector or mutating weights. The lower-level drift-corrector entry points also inspect the actual network.layerwise_offloader and reject all deployment modes before loading or capturing weight storage. corrector_mode="off" remains supported. Tests cover both application-level paths and the direct drift-corrector APIs.

**Comment by Shn0930** [comment]
@jarcherNV  Hi Jesse, can you help to review or assign the reviewer? Thank you!

**Comment by jarcherNV** [comment]
This looks like a reasonable change for reducing VRAM usage, at the cost of performance due to not having access to torch.compile or CUDA graphs, which is probably fine as long as this remains strictly opt-in.

@aidanfnv it looks like this modifies some of Crazy Taxi, does the change look ok to you?

**Comment by jarcherNV** [comment]
Also @gtong-nv and @ArielG-NV for visibility, and in case either of you want to review.

If we do want to take this change it might be worth thinking about broadening this to not just a feature for omni-dreams, but a more general flashdreams thing (ie trading torch.compile for lower memory usage).

**Comment by ArielG-NV** [comment]
`flashdreams/flashdreams/infra/acceleration/layerwise_offload.py` does not seem omnidreams specific. Looks like we would just need to add functionality to `StreamInferencePipeline`/`StreamInferencePipelineCache` to more generically support this feature (have not looked too deeply in implementation, so this may be wrong)?

**Comment by gtong-nv** [comment]
We have the `--offload-text-encoder` option omnidreams/interactive drive, and that reduces the peak VRAM usage quite lot and not hurting the performance given we only run the encoder in the beginning. 
But I'm not sure if the layer-wise DiT will work for the interactive usage.
The benchmark attached shows `median generate() latency by +0.154%`, which seems too good to be true. I'm going to run some benchmarks locally before we decide if we want to add this option.

**Comment by gtong-nv** [comment]
Benchmarked this locally on an RTX PRO 6000 Blackwell (96 GB), using the Interactive Drive configuration at 1280×704, one view, and eight output frames per steady-state chunk.

### Peak VRAM

| Configuration | Peak VRAM | Difference |
| --- | ---: | ---: |
| Resident | 25.03 GiB | — |
| Layer-wise offload | 21.98 GiB | -3.05 GiB (-12.2%) |

Layer-wise offload saves approximately 3 GiB out of a 25 GiB pipeline. This is useful, but is unlikely to be a major turning key for making the current model fit on substantially more GPU classes. The largest allocations are the resident text encoder (15.45 GiB) and self-attention KV cache (4.51 GiB), while the offloaded DiT blocks account for 3.61 GiB before GPU staging overhead.

### Throughput

| Configuration | Generate-only FPS | End-to-end FPS, including finalize |
| --- | ---: | ---: |
| Resident | 17.62 | 12.61 |
| Layer-wise offload | 16.11 | 11.46 |
| Change | -8.6% | -9.1% |

The FPS impact is smaller than expected. But feels no need for the current models and HWs. It may be more valuable for future models with larger weight footprints, where layer weights represent a greater share of total VRAM.

**Comment by Shn0930** [comment]
Has been adapted to support more models, instead of treating it as an exclusive feature of Omnidreams.
For the Interactive Drive pipeline, we load reasoner in VRAM, so that's the reason Dit is not the block. But in our users cases, they prefer fix the text embedding.
- **Cosmos-Reason1 text encoder: 15.45 GiB.** The wrapper loads the complete
  BF16 `Qwen2_5_VLForConditionalGeneration` checkpoint, including components
  that are not active in the text-only forward.
- **Self-attention rolling KV cache: 4.51 GiB.** At 1280x704, the latent grid is
  160x88 and the 2x2 DiT patch grid is 80x44. A six-frame window therefore has
  21,120 tokens. Across 28 blocks, two BF16 K/V tensors with 16 heads and
  128 values per head require 4.5117 GiB.
- **Offloadable DiT blocks: 3.61 GiB.** Offload removes most of this allocation
  but retains 0.2578 GiB of GPU staging slots.

For latency, I didn't do much test on Blackwell, I think rtx pro 6k compute power is much higher than H20/L4/L20. For some GPUs with less compute power, this computation and H2D transfer can overlap. I’m also trying to make further optimizations on communication. 
I also compare with CUDA Graph inference in H20. Here is the result.
| Mode | Median chunk (s) | Frames/s |
| --- | ---: | ---: |
| CUDA Graph, resident | 1.549386 | 5.163335 |
| Eager, resident | 1.554586 | 5.146063 |
| Eager, layer-wise offload | 1.556658 | 5.139215 |
